// @kiltprotocol/core/lib/cjs/balance/Balance.utils.js
"use strict";
/**
 * Copyright (c) 2018-2022, BOTLabs GmbH.
 *
 * This source code is licensed under the BSD 4-Clause "Original" license
 * found in the LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromFemtoKilt = exports.toFemtoKilt = exports.balanceNumberToString = exports.TRANSACTION_FEE = exports.convertToTxUnit = exports.formatKiltBalance = exports.Prefixes = exports.KILT_COIN = void 0;
const util_1 = require("@polkadot/util");
exports.KILT_COIN = new util_1.BN(1);
exports.Prefixes = new Map([
    ['femto', -15],
    ['pico', -12],
    ['nano', -9],
    ['micro', -6],
    ['milli', -3],
    ['centi', -2],
    ['KILT', 0],
    ['kilo', 3],
    ['mega', 6],
    ['mill', 6],
    ['giga', 9],
    ['bill', 9],
    ['tera', 12],
    ['tril', 12],
    ['peta', 15],
    ['exa', 18],
    ['zetta', 21],
    ['yotta', 24],
]);
/**
 * Uses the polkadot.js balance formatter, to convert given BN to a human-readable prefixed number.
 *
 * @param amount BN to format.
 * @param additionalOptions Optional formatting settings, these are defaulted to KILT specific settings.
 * @returns String representation of the given BN with prefix and unit ('KILT' as default).
 */
function formatKiltBalance(amount, additionalOptions) {
    const options = {
        withSi: true,
        decimals: 15,
        withSiFull: true,
        withUnit: 'KILT',
        ...additionalOptions,
    };
    return (0, util_1.formatBalance)(amount, options);
}
exports.formatKiltBalance = formatKiltBalance;
/**
 * Converts balance from KILT denomination to base unit.
 *
 * @param balance Balance in KILT denomination.
 * @param power Allows modifying conversion. Set to 0 for conversion to base unit, set to <0 for various larger denominations. -15 is KILT denomination.
 * @returns Converted (redenominated) balance.
 */
function convertToTxUnit(balance, power) {
    return new util_1.BN(balance).mul(new util_1.BN(10).pow(new util_1.BN(15 + power)));
}
exports.convertToTxUnit = convertToTxUnit;
exports.TRANSACTION_FEE = convertToTxUnit(new util_1.BN(125), -9);
/**
 * Safely converts the given [[BalanceNumber]] to a string, using the supplied methods,
 * or if given a string checks for valid number representation.
 *
 * @param input [[BalanceNumber]] to convert.
 * @returns String representation of the given [[BalanceNumber]].
 */
function balanceNumberToString(input) {
    if (typeof input === 'string') {
        if (!input.match(/^-?\d*\.?\d+$/)) {
            throw new Error('Not a string representation of number');
        }
        return input;
    }
    if (typeof input === 'number' ||
        typeof input === 'bigint' ||
        (typeof input === 'object' && input instanceof util_1.BN)) {
        return input.toString();
    }
    throw new Error('Could not convert to String');
}
exports.balanceNumberToString = balanceNumberToString;
/**
 * Converts the given [[BalanceNumber]] to the femto KILT equivalent.
 *
 * @param input [[BalanceNumber]] to convert.
 * @param unit Metric prefix of the given [[BalanceNumber]].
 * @returns Exact BN representation in femtoKilt, to use in transactions and calculations.
 */
function toFemtoKilt(input, unit = 'KILT') {
    const stringRepresentation = balanceNumberToString(input);
    if (!exports.Prefixes.has(unit)) {
        throw new Error('Unknown metric prefix');
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const unitVal = exports.Prefixes.get(unit);
    const negative = stringRepresentation.substring(0, 1) === '-';
    const [integer, fraction] = negative
        ? stringRepresentation.substring(1).split('.')
        : stringRepresentation.split('.');
    if (fraction && fraction.length > unitVal + 15) {
        throw new Error(`Too many decimal places: input with unit "${unit}" and value "${stringRepresentation}" exceeds the ${unitVal + 15} possible decimal places by ${fraction.length - unitVal + 15}`);
    }
    const fractionBN = fraction
        ? convertToTxUnit(new util_1.BN(fraction), unitVal - fraction.length)
        : new util_1.BN(0);
    const resultingBN = convertToTxUnit(new util_1.BN(integer), unitVal).add(fractionBN);
    return resultingBN.mul(new util_1.BN(negative ? -1 : 1));
}
exports.toFemtoKilt = toFemtoKilt;
/**
 * Converts the given [[BalanceNumber]] to a human-readable number with metric prefix and Unit.
 * This function uses the polkadot formatBalance function,
 * it's output can therefore be formatted via the polkadot formatting options.
 *
 * @param input [[BalanceNumber]] to convert from Femto Kilt.
 * @param decimals Set the minimum decimal places in the formatted localized output, default is 4.
 * @param options [[BalanceOptions]] for internationalization and formatting.
 * @returns String representation of the given [[BalanceNumber]] with unit und metric prefix.
 */
function fromFemtoKilt(input, decimals = 4, options) {
    const inputBN = new util_1.BN(balanceNumberToString(input));
    const formatted = formatKiltBalance(inputBN, options);
    const [number, ...rest] = formatted.split(' ');
    const negative = number.startsWith('-');
    const localeNumber = new Intl.NumberFormat(options?.locale, {
        minimumFractionDigits: 4,
    }).format(Number(negative ? number.slice(1) : number));
    return `${negative ? '-' : ''}${localeNumber.slice(0, localeNumber.length - 4 + decimals)} ${rest.join(' ')}`;
}
exports.fromFemtoKilt = fromFemtoKilt;
