// @kiltprotocol/core/lib/cjs/index.js
"use strict";
/**
 * Copyright (c) 2018-2022, BOTLabs GmbH.
 *
 * This source code is licensed under the BSD 4-Clause "Original" license
 * found in the LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDKErrors = exports.init = exports.disconnect = exports.connect = exports.Quote = exports.DelegationNodeUtils = exports.DelegationNode = exports.CType = exports.Claim = exports.BalanceUtils = exports.Credential = exports.Attestation = void 0;
const tslib_1 = require("tslib");
/**
 * @module @kiltprotocol/core
 */
exports.Attestation = tslib_1.__importStar(require("./attestation/index.js"));
exports.Credential = tslib_1.__importStar(require("./credential/index.js"));
var index_js_1 = require("./balance/index.js");
Object.defineProperty(exports, "BalanceUtils", { enumerable: true, get: function () { return index_js_1.BalanceUtils; } });
exports.Claim = tslib_1.__importStar(require("./claim/index.js"));
exports.CType = tslib_1.__importStar(require("./ctype/index.js"));
var index_js_2 = require("./delegation/index.js");
Object.defineProperty(exports, "DelegationNode", { enumerable: true, get: function () { return index_js_2.DelegationNode; } });
Object.defineProperty(exports, "DelegationNodeUtils", { enumerable: true, get: function () { return index_js_2.DelegationNodeUtils; } });
exports.Quote = tslib_1.__importStar(require("./quote/index.js"));
var index_js_3 = require("./kilt/index.js");
Object.defineProperty(exports, "connect", { enumerable: true, get: function () { return index_js_3.connect; } });
Object.defineProperty(exports, "disconnect", { enumerable: true, get: function () { return index_js_3.disconnect; } });
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return index_js_3.init; } });
var utils_1 = require("@kiltprotocol/utils");
Object.defineProperty(exports, "SDKErrors", { enumerable: true, get: function () { return utils_1.SDKErrors; } });
