// @kiltprotocol/sdk-js/lib/cjs/index.js
"use strict";
/**
 * Copyright (c) 2018-2022, BOTLabs GmbH.
 *
 * This source code is licensed under the BSD 4-Clause "Original" license
 * found in the LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = exports.Did = exports.ChainHelpers = exports.Blockchain = exports.Message = exports.ConfigService = void 0;
const tslib_1 = require("tslib");
/**
 * @module @kiltprotocol/sdk-js
 */
tslib_1.__exportStar(require("@kiltprotocol/core"), exports);
var config_1 = require("@kiltprotocol/config");
Object.defineProperty(exports, "ConfigService", { enumerable: true, get: function () { return config_1.ConfigService; } });
exports.Message = tslib_1.__importStar(require("@kiltprotocol/messaging"));
var chain_helpers_1 = require("@kiltprotocol/chain-helpers");
Object.defineProperty(exports, "Blockchain", { enumerable: true, get: function () { return chain_helpers_1.Blockchain; } });
exports.ChainHelpers = tslib_1.__importStar(require("@kiltprotocol/chain-helpers"));
exports.Did = tslib_1.__importStar(require("@kiltprotocol/did"));
exports.Utils = tslib_1.__importStar(require("@kiltprotocol/utils"));
tslib_1.__exportStar(require("@kiltprotocol/types"), exports);
